brave/RATIONALE.md

```diff
@@ -228,6 +228,101 @@ implementations proposed in these papers are different to the implementation
+propagation. There are some subtle points easy to miss about the design.
+
+* Decode is called before a trace context exists, so there's no `context` parameter.
+  * First successful decode returns `true`.
+    * This allows us to prioritize and leniently read, multiple header names.
+  * The `ValueUpdater` parameter allows us to hide or mute context storage
+    * We considered `Propagation.Setter`, but it doesn't return boolean
+  * The `request` parameter allows secondary sampling to use `SamplerFunction` on inbound requests.
+  * The `value` parameter could be a delimited string, or a plain value to set.
+
+* Encode is called only when a baggage exists and isn't all redacted
+  * The `Map<String, String>` parameter allows implementations to encode all
+    baggage in one string.
+  * The `context` parameter allows `BaggageField.getValue` in simple case (more
+    efficient than map).
+    * The `context` parameter also allows secondary sampling to write down the
+      `spanId`
+  * The `request` parameter allows secondary sampling to use `SamplerFunction`
+    on outbound requests.
+
+#### Why does decode not accept a Map?
+When looking at the encode vs decode side, it might seem curious why they don't
+both implement `Map`. `Map` is definitely needed on the encode side, as it
+supports encoding all baggage in a single header. While decoding one header can
+result in multiple baggage fields, the code to accomplish that is as easy with
+a map as a functional interface. It is also easier for the baggage code to
+implement a function over `BaggageField.updateValue` vs try to implement the
+entire write side of `Map`.
+
+### `Map` views over `BaggageField` values
+As `Map` is a standard type, it is a safer type to expose to consumers of all
+baggage fields. We decided to use an unmodifiable `Map` instead of an internal
+type for use cases such as coalescing all baggage into a single header. This
+map must also hide local fields from propagation.
+
+#### Why not a standard `Map` type
+We considered copying the internal state array to an existing `Map` type, such
+as `LinkedHashMap`. However, doing so would add overhead regardless of it that
+`Map` was ever used, or if that map had a value for the field the consumer was
+interested in! Concretely, we'd pay to create the map, to copy the values in
+the array into it, and also pay implicit cost of iterator allocation for
+operations such as `entrySet()` even if the entry desired was not present. This
+holds present even extending `AbstractMap`, as most operations, even `get()`
+allocate iterators.
+
+#### `UnsafeArrayMap`
+We created `UnsafeArrayMap` as a view over our pair-wise internal state array.
+The implementation is very simple except that we have a redaction use case to
+address, which implies filtering keys. To address that, we keep a bitmap of all
+filtered keys and consider that when performing any scan operations.
```

Comment:
This also was just moved (I moved them as I noticed we already had a baggage heading!